algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs

```diff
@@ -21,249 +20,286 @@ use crate::{
-                    &circuit.ifft_precomputation,
-                );
-                (*circuit, result)
-            });
-
-            let label_g_c = witness_label(circuit.id, "g_c", 0);
-            pool.add_job(move || {
-                let result = Self::matrix_sumcheck_helper(
-                    label_g_c,
-                    circuit_state.non_zero_c_domain,
-                    &circuit.c_arith,
-                    *alpha,
-                    beta,
-                    v_H_i_alpha_v_H_i_beta,
-                    largest_non_zero_domain_size,
-                    &circuit.fft_precomputation,
-                    &circuit.ifft_precomputation,
-                );
-                (*circuit, result)
-            });
-        }
+        let zk_bound = Self::zk_bound();
 
-        let mut sums = Vec::with_capacity(state.circuit_specific_states.len());
-        let mut gs = BTreeMap::new();
-        for ((circuit_a, (sum_a, lhs_a, g_a)), (circuit_b, (sum_b, lhs_b, g_b)), (circuit_c, (sum_c, lhs_c, g_c))) in
-            pool.execute_all().into_iter().tuples()
-        {
-            assert_eq!(circuit_a, circuit_b);
-            assert_eq!(circuit_a, circuit_c);
-            let matrix_sum = prover::message::MatrixSums { sum_a, sum_b, sum_c };
-            sums.push(matrix_sum);
-            state.circuit_specific_states.get_mut(circuit_a).unwrap().lhs_polynomials = Some([lhs_a, lhs_b, lhs_c]);
-            let matrix_gs = prover::MatrixGs { g_a, g_b, g_c };
-            gs.insert(circuit_a.id, matrix_gs);
-        }
+        let max_variable_domain = state.max_variable_domain;
+
+        let verifier::FirstMessage { batch_combiners } = verifier_message;
+        let verifier::SecondMessage { alpha, eta_b, eta_c } = verifier_second_message;
 
-        let msg = prover::ThirdMessage { sums };
-        let oracles = prover::ThirdOracles { gs };
+        let (h_1, x_g_1_sum, msg) = Self::calculate_lhs(&mut state, batch_combiners, alpha, eta_b, eta_c)?;
 
-        assert!(oracles.matches_info(&Self::third_round_polynomial_info(
-            state.circuit_specific_states.keys().map(|c| (c.id, &c.index_info))
-        )));
+        let mut sumcheck_lhs = h_1.mul_by_vanishing_poly(max_variable_domain);
+        sumcheck_lhs += &x_g_1_sum;
+        debug_assert!(
+            sumcheck_lhs.evaluate_over_domain_by_ref(max_variable_domain).evaluations.into_iter().sum::<F>()
+                == msg.sum(batch_combiners, *eta_b, *eta_c)
+        );
+
+        let g_1 = DensePolynomial::from_coefficients_slice(&x_g_1_sum.coeffs[1..]);
+
+        drop(x_g_1_sum);
+
+        assert!(g_1.degree() <= max_variable_domain.size() - 2);
+        assert!(h_1.degree() <= 2 * max_variable_domain.size() + 2 * zk_bound.unwrap_or(0) - 2);
+
+        let oracles = prover::ThirdOracles {
+            g_1: LabeledPolynomial::new("g_1".into(), g_1, Some(max_variable_domain.size() - 2), zk_bound),
+            h_1: LabeledPolynomial::new("h_1".into(), h_1, None, None),
```

Comment:
Maybe we can change the `LabeledPolynomial::new` API to `fn new(label: impl Into<String>, polynomial: DensePoly, degree_bound: impl Into<Option<usize>>, hiding_bound: impl Into<Option<usize>>)` so that we don't have to manually call `into()` and annotate with `Some`

eg the above two lines would then become:
```rust
g_1: LabeledPolynomial::new("g_1", g_1, max_variable_domain_size() - 2, zk_bound),
h_1: LabeledPolynomial::new("h_1", g_1, None, None)
```

Reply:
601f8ec03